os/arch/arm/src/s5j/s5j_i2s.c

```diff
@@ -2011,26 +2026,81 @@ static int i2s_irq_handler(int irq, FAR void *context, FAR void *arg)
 
-	if (reg & I2S_CR_FTXURSTATUS) {
-		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FTXURINTEN, 0);
-		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_FTXURSTATUS);
-		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_TXDMACTIVE, 0);
-		if (priv->err_cb)
-			priv->err_cb((struct i2s_dev_s *)priv, priv->err_cb_arg, I2S_ERR_CB_TX);
-	
+#ifdef I2S_HAVE_RX
+	/* This is the case, when there is a new transfer series 
+	 * is initiated right before RX FIFO OVF 
+	 * It can be a new series or delay/lag in users application receiver.
+	*/	
+
+	if (!sq_empty(&priv->rx.pend) && sq_empty(&priv->rx.act)) {
+		/* Remove the next buffer container from the tx.pend list */
+		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.pend);
+
+		/* Add the completed buffer container to the tail of the tx.act queue */
+		sq_addlast((sq_entry_t *) bfcontainer, &priv->rx.act);
+
+		/* Start next DMA transfer */
+		s5j_dmastart(priv->rx.dma, bfcontainer->dmatask);
+		
+		/* We don't stop I2S operation, report anyways about OVF */
+		if (reg & I2S_CR_FRXOFSTATUS) {
+			if (priv->err_cb) {
+				priv->err_cb((struct i2s_dev_s *)priv, priv->err_cb_arg, I2S_ERR_CB_RX);
+			}
+		}
+	} else {
+		/* If OVF we report */
+		if (reg & I2S_CR_FRXOFSTATUS) {
+			modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FRXOFINTEN, 0);
+			modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_FRXOFSTATUS);
+			modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_RXDMACTIVE, 0);
+			if (priv->err_cb) {
+				priv->err_cb((struct i2s_dev_s *)priv, priv->err_cb_arg, I2S_ERR_CB_RX);
+			}
+		}
 	}
+#endif
+
+#ifdef I2S_HAVE_TX_P
+	/* This is the case, when there is a new transfer series 
+	 * is initiated right before TX FIFO UNF 
+	 * It can be a new series or delay/lag in users application receiver.
+	*/	
+	if (!sq_empty(&priv->txp.pend) && sq_empty(&priv->txp.act)) {
+		/* Remove the next buffer container from the tx.pend list */
+		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.pend);
+
+		/* Add the completed buffer container to the tail of the tx.act queue */
+		sq_addlast((sq_entry_t *) bfcontainer, &priv->txp.act);
+
+		/* Start next DMA transfer */
+		s5j_dmastart(priv->txp.dma, bfcontainer->dmatask);
 
-	if (reg & I2S_CR_FRXOFSTATUS) {
-		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FRXOFINTEN, 0);
-		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_FRXOFSTATUS);
-		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_RXDMACTIVE, 0);
-		if (priv->err_cb)
-			priv->err_cb((struct i2s_dev_s *)priv, priv->err_cb_arg, I2S_ERR_CB_RX);
+	
+		/* We don't stop I2S operation, report anyways about UNF */
+		if (reg & I2S_CR_FTXURSTATUS) {
+			if (priv->err_cb) {
+				priv->err_cb((struct i2s_dev_s *)priv, priv->err_cb_arg, I2S_ERR_CB_TX);
+			}
+		}
+	} else {
+		/* If UNF we report */
+		if (reg & I2S_CR_FTXURSTATUS) {
+			modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FTXURINTEN, 0);
+			modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_FTXURSTATUS);
+			modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_TXDMACTIVE, 0);
```

Comment:
Above code stops and disables TX channel.
When next TX data will be submitted, TX channel will be restarted.